### PR TITLE
(842) Reason validation

### PR DESCRIPTION
**Pull request by jsrobertson**
## Context

https://trello.com/c/OkMpaWcU/842-make-reason-for-referral-required-ui

We have decided to create a new controller for each step of a referral to prevent the possibility of the `ReferralsController` becoming too unwieldy.

## Changes in this PR
This extracts the logic for showing, setting and updating the `reason` field on a Referral out of the ReferralsController and into an `ReasonController`.

- Add new `ReasonController`
- Move reason related method and functionality from ReferralsController to new `ReasonController`
- Add new `update` reason method 
- Display error messaging on reason page when no reason submitted

## Screenshots of UI changes

localhost_3000_referrals_a16f58c0-1c2a-495e-ae8d-823e5e9518c5_reason

## Release checklist

[Release process documentation](../doc/how-to/perform-a-release.md)

As part of our continuous deployment strategy we must ensure that this work is
ready to be released once merged.

### Pre-merge

- [ ] There are changes required to the Accredited Programmes API for this change to work...
  - [ ] ... and they have been released to production already

### Post-merge

- [ ] This adds, extends or meaningfully modifies a feature...
  - [ ] ... and I have written or updated an end-to-end test for the happy path in the [Accredited Programmes E2E repo](https://github.com/ministryofjustice/hmpps-accredited-programmes-e2e)
- [ ] This makes new expectations of the API...
  - [ ] ... and I have notified the API developer(s) of changes to the contract tests (Pact), or the API is already compliant
- [ ] [Manually approve](../doc/how-to/perform-a-release.md#releasing-to-the-preprod-environment) release to preprod
- [ ] [Manually approve](../doc/how-to/perform-a-release.md#releasing-to-the-production-environment) release to prod

### CODE REVIEW

File: server/controllers/refer/reasonController.ts [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+import { FormUtils, TypeUtils } from '../../utils'
+import type { ReferralUpdate } from '@accredited-programmes/models'
+
+export default class ReasonController {
+  constructor(
+    private readonly personService: PersonService,
+    private readonly referralService: ReferralService,
+  ) {}
+
+  show(): TypedRequestHandler<Request, Response> {
```
Comment (jsrobertson) [inline server/controllers/refer/reasonController.ts:15]: Probably a good opportunity to start specifying the Params and Request Body types. Will look into it next week.

Comment (yndajas) [reply server/controllers/refer/reasonController.ts:15]: That sounds good - maybe as a separate PR?

File: server/views/referrals/reason.njk [hunk 2]
```diff
@@ -19,6 +20,12 @@
 {% block content %}
   <div class="govuk-grid-row">
     <div class="govuk-grid-column-two-thirds">
+      {% if errors.list.length %}
+        {{ govukErrorSummary({
+          titleText: "There is a problem",
+          errorList: errors.list
+        }) }}
+      {% endif %}
```
Comment (yndajas) [inline server/views/referrals/reason.njk:28]: I wonder if this should be two commits - one to start using the new controller action, and then one to display the errors (both view and controller changes)

**Review by yndajas — APPROVED** [review]
Looks good. Just a few things that might be good to tidy up

File: server/controllers/refer/reasonController.test.ts [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+      await requestHandler(request, response, next)
+
+      expect(referralService.updateReferral).toHaveBeenCalledWith(token, referral.id, {
+        oasysConfirmed: referral.oasysConfirmed,
+        reason: 'Some reason\nAnother paragraph',
+      })
+      expect(response.redirect).toHaveBeenCalledWith(`/referrals/${referral.id}`)
+    })
+
+    describe('when the reason is not provided', () => {
```
Comment (yndajas) [inline server/controllers/refer/reasonController.test.ts:99]: Might be good to also test providing a reason that's just space characters (`  ` or ` \n  `) - not sure if it's worth integration testing this too, but probably not?

**Pull request merged**